src/storage/buffer_manager/buffer_manager.cpp

```diff
@@ -13,35 +11,32 @@ namespace storage {
 // eviction queue until we hit a candidate that is evictable.
-// Two kinds of candidates are not evictable: 1) it is currently pinned; 2) it has been recently
-// visited.
-// To identify those recently accessed candidates, we use the eviction timestamp. If the
-// eviction timestamp of a candidate is different from the timestamp in its corresponding pageState,
-// it means that the candidate has been recently visited and we should not evict it. The idea is
-// that eviction timestamp is a logical per-page timestamp starting from 0, and is incremented each
-// time the page is pushed into the eviction queue. For example, the first time p5 is pushed into
-// the eviction queue, it will end up with a timestamp 1. When we push a page into the queue, we
-// create an EvictionCandidate object for the page. Let's call this object c0 when p5 is first
-// pushed. c0 will consist of (ptr to p5, 1), where the latter is the eviction timestamp at the time
-// c0 is put into the queue. Suppose p5 is put into the eviction queue again (e.g., because it was
-// pinned and unpinned). At this point we create another EvictionCandidate object c1 (ptr to p5, 2)
-// where the latter eviction timestamp is now incremented by 1, which makes c1 now not evictable.
-// This idea is inspired by DuckDB's queue-based eviction implementation.
+// 1) If the candidate page's version has changed, which means the page was pinned and unpinned, we
+// remove the candidate.
+// 2) If the candidate page's state is UNLOCKED, and its page version hasn't changed, which means
+// the page was optimistically read, we mark it as MARKED, and move it to the back of the queue.
+// 3) If the candidate page's state is LOCKED, we remove the candidate.
 void EvictionQueue::removeNonEvictableCandidates() {
     EvictionCandidate evictionCandidate;
     while (true) {
         if (!queue->try_dequeue(evictionCandidate)) {
             break;
         }
-        if (evictionCandidate.pageState->getPinCount() != 0 ||
-            evictionCandidate.pageState->getEvictionTimestamp() !=
-                evictionCandidate.evictionTimestamp) {
-            // Remove the candidate from the eviction queue if it is still pinned or if it has
-            // been recently visited.
+        auto pageStateAndVersion = evictionCandidate.pageState->getStateAndVersion();
+        if (evictionCandidate.isSecondChanceEvictable(pageStateAndVersion)) {
+            // The page was optimistically read, mark it as MARKED, and enqueue to be evicted later.
+            evictionCandidate.pageState->tryMark(pageStateAndVersion);
+            queue->enqueue(evictionCandidate);
             continue;
-        } else {
+        } else if (evictionCandidate.isEvictable(pageStateAndVersion)) {
             queue->enqueue(evictionCandidate);
             break;
+        } else {
+            // Cases to remove the candidate from the queue:
+            // 1) The page is currently pinned, remove the candidate from the queue.
```

Comment:
I would clarify the comment to say talk about the state first instead of the action that lead to the state. So something like: 
`The page is currently LOCKED (because it was pinned), remove ...`
`The page's version number was different (because it was pinned and unpinned ...`.

Reply:
Done.